File: julia/2021/day06.jl

function simulate(fish_cycle_count::AbstractVector, epochs::Int)
    for _ = 1:epochs
        cycle_end = popfirst!(fish_cycle_count)
        push!(fish_cycle_count, cycle_end)
        fish_cycle_count[6+1] += cycle_end
    end
    return sum(fish_cycle_count)
end

function main()
    # input = parse.(Int, split("3,4,3,1,2", ','))
    input = parse.(Int, split(readline(joinpath(@__DIR__, "input/day06.txt")), ','))
    init_cycle_count = [count(==(i), input) for i = 0:8]

    part1 = simulate(copy(init_cycle_count), 80)
    part2 = simulate(copy(init_cycle_count), 256)
    return part1, part2
end

println(main())

File: 2021/javascript/src/day06.js
const { readFileSync } = require("fs");

function simulate(fish_cycle_count, epochs) {
  for (let i = 0; i < epochs; i++) {
    let cycle_end = fish_cycle_count.shift();
    fish_cycle_count.push(cycle_end);
    fish_cycle_count[6] += cycle_end;
  }
  return fish_cycle_count.reduce((a, b) => a + b, 0);
}

function day06(fileName = "./data/day06.txt") {
  const input = readFileSync(fileName, "utf8")
    .split(",")
    .map((i) => parseInt(i));

  // input = [3, 4, 3, 1, 2];
  const init_cycle_count = [0, 1, 2, 3, 4, 5, 6, 7, 8].map(
    (n) => input.filter((i) => i === n).length
  );

  const part1 = simulate([...init_cycle_count], 80);
  const part2 = simulate([...init_cycle_count], 256);

  return [part1, part2];
}

module.exports = {
  day06,
};
